### DISCUSSION TITLE
Copilot Extensions now in Public Beta

### DISCUSSION THREAD

**Discussion by LiteBrite82** [community/community#137975] [Copilot Conversations]
# Copilot Extensions now in Public Beta

![Image](https://github.com/user-attachments/assets/2941f169-4f9f-4bd3-9f8f-66ec7152a983)

We’re thrilled to announce the public beta release of GitHub Copilot Extensions, available starting September 9th, 2024. Whether you’re a developer looking to leverage existing extensions or a builder excited to create new integrations, Copilot Extensions now offers a platform to make that happen. Copilot Extensions allow developers to build or install integrations into Copilot Chat on GitHub.com and all supported editors (Visual Studio and VS Code) to increase productivity by seamlessly integrating external tools and context into the developer workflow.

### What's New in the Public Beta?

- **Open Access**: All Copilot users, organizations, and enterprises (on supported plans) can now install and use Copilot Extensions.
- **Build your own extension**: Any developer or organization can create Copilot extensions, including private and public extensions for organizations, and the option to publish your extension to the GitHub Marketplace
- **Extension Toolkit**: We've released a comprehensive set of resources for builders; with a learning pathway, [docs](https://gh.io/getting-started-copilot-extension-building), the [Copilot-Extensions](https://github.com/orgs/copilot-extensions/) organization with code samples and an initial SDK, and a [CLI debugging tool](https://github.com/copilot-extensions/gh-debug-cli).

### Get Started Today
Explore the [GitHub Marketplace](https://github.com/marketplace?type=apps&copilot_app=true) to try a Copilot Extension, or check out our [Getting Started Guide](https://gh.io/getting-started-copilot-extension-building) to build your own.

### Feedback
We’re eager to hear about your experience, as a user or builder, during pubic beta.

For general feedback or feature requests around using and building extensions, please post in the comments below.
For detailed builder feedback to the GitHub product team, fill out the [Extension Builders Survey](https://survey3.medallia.com/?Mxu02i-extensibility).
For user feedback to a specific extension publisher, please file an issue in the [user-feedback repo](https://github.com/copilot-extensions/user-feedback) with a label for the publisher's listing name attached.

**Comment by kyleli626** [comment]
Feature request - 

Chat participants, and thus copilot (and copilot chat) extensions, don’t seem to work for inline-chat
- [possible pr](https://github.com/microsoft/vscode/pull/209664)
- [closed issue](https://github.com/microsoft/vscode/issues/208898)

For example, try @ workspace in an inline chat. This seems to limit ability of extensions, especially since (in my experience), I tend to use more inline chat vs chat view.

**Reply by sharonlo** [reply]
That's great feedback @kyleli626! definitely looking into how we support extensions to inline chat :)

**Comment by Tolga1452** [comment]
I have a question (for the future versions of this beta). Will GitHub send references (such as selected code) to the endpoint along with the history, just like how GitHub Copilot works? Because I need to make a VS Code extension to get the code references from the IDE, but if I do that, the extension adds another participant to the chat and probably I need to implement a separate system within the extension, which would differ from the one my server currently uses to handle requests from GitHub, especially because I get different data from the VS Code API.

I also wonder if other chat extension capabilities will be added. Because I would like to send custom buttons, follow-up messages etc. to the chat. And the most disappointing part is I expected to create commands for my Copilot Extension without having to build an IDE extension.

I would like to have these features because with the current usage of Copilot Extensions (I also took a look at the [dreamcode](https://github.com/copilot-extensions/preview-sdk.js/blob/main/dreamcode.md)), we can only provide a chat experience to the users, which is far away from helping to code.

**Reply by D1M1TR10S** [reply]
Hi @Tolga1452!
> Will GitHub send references (such as selected code) to the endpoint...

Currently, "references" within the scope of extensibility are sent. We are going to support additional context, like selected code, within the next two months. This will apply specifically to Copilot Extensions built as GitHub Apps. For VS Code Chat Participants, selected code is available as context. You can refer to the VS Code documentation for details on VS Code Chat specific capabilities.

> I also wonder if other chat extension capabilities will be added...?

Capabilities like custom buttons and follow-up messages are common requests that we're tracking. While these aren't on the immediate roadmap, we are considering these features for future updates.

We understand that a lot of builders want more capabilities beyond chat, and we’re actively exploring ways to support deeper integration as we continue building out the platform.

**Reply by Tolga1452** [reply]
Thanks! ❤️ I'm looking forward to the future updates

**Comment by abdessalems** [comment]
❤️

**Comment by abdessalems** [comment]
❤️

**Comment by abdessalems** [comment]
❤️

**Comment by dipuharidasan** [comment]
Thank you for the public beta - it has great potential! I was able to build a simple chat participant VS Code extension using [this guide](https://code.visualstudio.com/api/extension-guides/chat), which sends user queries to our internal LLM and posts the API responses back to the Copilot Chat window.

In this scenario, does Copilot read the data from our internal LLM that is posted back to the chat? If so, will this data be treated the same as the prompts and suggestions mentioned in the [Trust Center](https://copilot.github.trust.page/faq#security)? I am a GH Enterprise user.

**Reply by rajbos** [reply]
As the entire chat is send back and forth, GitHub will see the response you send back to the client. So yes. 

Also be aware that the response is also send to Azure OpenAI for the content and safety filtering.

**Reply by dipuharidasan** [reply]
Thanks, Rob. I assume that the same would apply to any Copilot Chat responses, and the same data privacy policies would be applicable in this case as well.

**Comment by htekdev** [comment]
Are we able to have private extensions created only for our organization organization?

**Reply by D1M1TR10S** [reply]
Hey @htekdev. Yes! You can create private extensions at the organization level.

The process will be the same as GitHub Apps (here's the link to our [docs on how to do that](https://docs.github.com/en/apps/creating-github-apps/registering-a-github-app/making-a-github-app-public-or-private)).

Currently, we don't support internal enterprise-level apps or extensions that can be shared across multiple organizations under the same enterprise. But this is on our roadmap and scheduled to release during Universe at the end of this month.

**Comment by isidorn** [comment]
Hi VS Code PM here 👋 

Just wanted to add to the initial announcement. You can also build Copilot Extensions in VS Code. This is an interesting path if you already have a VS Code extension. More details about this can be found here
https://code.visualstudio.com/docs/copilot/copilot-extensibility-overview
https://code.visualstudio.com/api/extension-guides/chat

Thanks 🙏

**Comment by yne** [comment]
Hi VS Code extension dev here.

My current company only use on-premise LLM, and I was hopping to integrate them in VScode using this lm and chat APIs but I got a little confused using them, could you clarify if

- Is it possible to use the VScode `lm` and `chat` APIs without a Copilot subscription?
  - If that possible: which extension samples showcase it ? (I tried 3 samples, all required copilot)
  - If that's not possible: why are they part of the VScode API and not in a copilot-specific repo ?

While testing, I triggered some Vscode-native chat dialogs (ex: "workbench.action.openQuickChat") yet my test extension was not called after entering some prompt.

**Reply by isidorn** [reply]
Hi there 👋 

> Is it possible to use the VScode lm and chat APIs without a Copilot subscription?

Right now no. But these APIs are designed to be generic, and this might change in the future depending on user feedback and potential scenarios.

> why are they part of the VScode API and not in a copilot-specific repo

Design decision we did that we believe makes it easy to extend Copilot on the VS Code side.

> While testing, I triggered some Vscode-native chat dialogs (ex: "workbench.action.openQuickChat") yet my test extension was not called after entering some prompt.

I suggest to do these two tutorials to help get started
https://code.visualstudio.com/api/extension-guides/chat-tutorial
https://code.visualstudio.com/api/extension-guides/language-model-tutorial

**Comment by biosugar0** [comment]
How Can I Create a Private Extension at the Organization Level?

I followed the steps in [this documentation](https://docs.github.com/en/copilot/building-copilot-extensions/managing-the-availability-of-your-copilot-extension) to create a private GitHub App for my organization, configured the extension, and installed it in the organization. However, the option for Copilot Chat does not appear with the @ prefix.

GitHub Copilot Extension is enabled for my organization, but it does not show up in the options. Interestingly, when I follow the same steps for a personal GitHub App setup, the option appears and works correctly.

For testing, I used https://github.com/copilot-extensions/blackbeard-extension, ran it locally, and set up the generated ngrok URL as the endpoint.

**Reply by isidorn** [reply]
@D1M1TR10S should be able to help

**Comment by kadirtugarantibbva** [comment]
Hello,

I believe github Copilot Extension feature is not available for **Copilot Only** Subscriptions for Enterprises. We don't have Github Enterprise accounts . We are not able to Enable the option for copilot extensions from policies. We also can't deploy Github apps or create repositories since we have copilot only accounts.

Is there a roadmap to serve this feature for also **Copilot Only** accounts? This will be a huge leap for us if we have this in our organization.

**Reply by hmorneau** [reply]
It's available to everyone with a GH Copilot subscription now.

**Comment by pidgeon777** [comment]
I'm encountering an issue with the [Stack Overflow extension](https://github.com/marketplace/stack-overflow-extension-for-github-copilot).

I added it to my account, but whenever I try to perform a request to it, my IDE reports an error (empty response).

This is very strange though, because it seems it has the necessary permissions:

![image](https://github.com/user-attachments/assets/87f71186-aa6f-43a1-a0fb-68ccf24321b4)

[Perplexity AI estension](https://github.com/marketplace/perplexityai) is working on my account, though. When I ask for some info about GitHub Copilot extensions, I get:

```
GitHub Copilot Extensions enhance the capabilities of GitHub Copilot by integrating external tools directly into the Copilot Chat experience. These extensions allow developers to interact with various services using natural language, significantly reducing context switching and streamlining workflows. Key features include the ability to query documentation, retrieve data, and execute actions across supported platforms such as Visual Studio Code, GitHub.com, and JetBrains IDEs. Organizations can create private extensions for internal use or share public extensions on the GitHub Marketplace. Notable partners contributing to this ecosystem include DataStax, Docker, and Microsoft Azure, among others. The aim is to empower developers to build and innovate more efficiently by leveraging a wide array of tools within a single interface.

**`References:`**
[perplexityai](https://github.com/marketplace/perplexityai)
[Introducing GitHub Copilot Extensions: Unlocking unlimited possibilities ...](https://github.blog/news-insights/product-news/introducing-github-copilot-extensions/)
[What is GitHub Copilot?](https://docs.github.com/en/copilot/about-github-copilot/what-is-github-copilot)
[Using extensions to integrate external tools with Copilot Chat - GitHub Docs](https://docs.github.com/en/copilot/using-github-copilot/using-extensions-to-integrate-external-tools-with-copilot-chat)
[About GitHub Copilot Completions in Visual Studio - Microsoft Learn](https://learn.microsoft.com/th-th/visualstudio/ide/visual-studio-github-copilot-extension?view=vs-2022&viewFallbackFrom=vs-2019)
[About building Copilot Extensions - GitHub Enterprise Cloud Docs](https://docs.github.com/en/enterprise-cloud@latest/copilot/building-copilot-extensions/about-building-copilot-extensions)
[Enhancing the GitHub Copilot ecosystem with Copilot Extensions ...](https://github.blog/news-insights/product-news/enhancing-the-github-copilot-ecosystem-with-copilot-extensions-now-in-public-beta/)
[About building Copilot Extensions - GitHub Docs](https://docs.github.com/en/copilot/building-copilot-extensions/about-building-copilot-extensions)
[GitHub Copilot extension - Overview - Azure Data Studio - Microsoft Learn](https://learn.microsoft.com/en-us/azure-data-studio/extensions/github-copilot-extension-overview)
```

Any help?

**Reply by pidgeon777** [reply]
I revoked the Stack Overflow authorization, then I performed again a request to the Stack Overflow extension. I got this answer:

```
Failed to authenticate. Visit following url to authorize askstackoverflow:
https://ide-extension-api.stackoverflow.net/login/github-copilot
```

Clicking on that link led me to:

![image](https://github.com/user-attachments/assets/de4e6cef-5cfa-44ef-84ae-f8fbe120d724)

I clicked on "Continue", but then a page loaded with this TXT answer (no HTML):

```
API access is not supported on this channel
```

Could my issue be related to that? Why it is happening? I own a regular Stack Overflow account.

Also, why that [Stack Overflow Teams](https://stackoverflowteams.com) website was opened instead of the "regular" one?

**Reply by hmorneau** [reply]
I have the same issue. It seems like API access to stackoverflow is only for paid account with them. But no where in the GH Copilot Marketplace they mention anything about that. This need to be reviewed, as it gives a cryptic error and make people believe that this extension is free to use which is misleading.

**Reply by pidgeon777** [reply]
> It seems like API access to stackoverflow is only for paid account with them.

But @hmorneau I suppose this is your hypothesis?

**Reply by hmorneau** [reply]
> But @hmorneau I suppose this is your hypothesis?

No, that's according to their documentation: "The Stack Overflow application programming interface (API) allows you to access your Basic and Business Teams data for automation, reporting, and more. Business Teams can use the API to read and write Teams data, while Basic Teams have read-only API access. **Free Teams do not have API access.**"

Ref: https://stackoverflowteams.help/en/articles/4385859-stack-overflow-for-teams-api

**Reply by pidgeon777** [reply]
I am unsure of this StackOverflow extension's usefulness. Initially, I believed it would serve as a search engine to access and summarize StackOverflow posts.

**Comment by hmorneau** [comment]
I find the way to find extension confusing. In the copilot chat inside VS Code when you invoke @ you have the option to install chat extensions which search the market place with this tag: @tag:chat-participant
<img width="445" alt="image" src="https://github.com/user-attachments/assets/1a58fbbc-03b9-4e75-b40a-4e230994ed29" />
Then you have other chat extension that do not show up with this tag like the pg extension:
<img width="548" alt="image" src="https://github.com/user-attachments/assets/8d5da79f-a4e9-4228-ab08-cc84ee5e60f9" />
Then you have a market place for extension as well: https://github.com/marketplace?type=apps&copilot_app=true
